overlord/snapstate/cookies.go

```diff
@@ -0,0 +1,137 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// GenerateCookies creates snap cookies for snaps that are missing them (may be the case for snaps installed
+// before the feature of running snapctl outside of hooks was introduced, leading to a warning
+// from snap-confine).
+// It is the caller's responsibility to lock state before calling this function.
+func (m *SnapManager) GenerateCookies(st *state.State) error {
+	var snapNames map[string]*json.RawMessage
+	if err := st.Get("snaps", &snapNames); err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+
+	snapWithCookies := make(map[string]bool)
+	for _, snap := range snapCookies {
+		snapWithCookies[snap] = true
+	}
+
+	var changed bool
+	for snap := range snapNames {
+		if _, ok := snapWithCookies[snap]; !ok {
+			cookieID, err := createCookieFile(snap)
+			if err != nil {
+				return err
+			}
+
+			snapCookies[cookieID] = snap
+			changed = true
+		}
+	}
+
+	if changed {
+		st.Set("snap-cookies", &snapCookies)
+	}
+	return nil
+}
+
+func (m *SnapManager) createSnapCookie(st *state.State, snapName string) error {
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+
+	// make sure we don't create cookie if it already exists
```

Comment:
O(N) complexity

Reply:
Sure. But N is the number of installed snaps, so a couple of hundreds in the worst case, and it's only executed in very specific cases (when you do ops on snaps such as install/uninstall). I'd argue that the cost here is negligible compared to all the I/O associated with the main operation (including state.json updates).